Alternate email address processing [TASK] #12

arsalaerfan opened the pull request:
## Alternate email address processing

This PR adds the endpoint for handling alternate email address submissions. Students can modify the email addresses registered under their account and designate their primary email for CTI communications.

### Changes Included
Implemented `POST /api/students/alternate-emails` endpoint.
* Add Alternate Emails – Allows students to add new emails, ensuring they aren’t linked to another student.
* Remove Alternate Emails – Deletes emails from a student’s record.
* Set Primary Email – Updates the primary email if its similar to the google form email.
* Enforce Email Uniqueness – Ensures each email belongs to only one student, rejecting duplicates.
* Prevent Primary Email Removal Without Reassignment – Blocks removal of the primary email unless a new one is set first.
* Authenticate Using Google Form Email – Identifies students via their submitted Google Form email

### Issues Fixed
* https://github.com/NickGuerrero/cti-sys/issues/12

### Tests

`TestModifyAlternateEmails` pytest class includes:
* 5 success tests (adding, removing, setting primary).
* 5 failure tests (conflicting emails, removing primary, invalid primary).

To run tests locally:
```
pytest -v -k "TestModifyAlternateEmails"
```

<img width="1028" alt="Screenshot 2025-03-08 at 3 58 53 PM" src="https://github.com/user-attachments/assets/6d6329fa-7b5b-4e44-a9cb-a7f410f82486" />


On Github Actions:

```
pytest -v -m "not integration"
```
<img width="1379" alt="Screenshot 2025-03-08 at 3 59 56 PM" src="https://github.com/user-attachments/assets/5dbca635-a04c-4544-a56f-3d1a48bd67e2" />


### Authentication Consideration
* Currently, authentication is handled by using the email submitted via the Google Form to identify the student.

### Notes on Library changes
* psycopg-binary version 3.2.3 no longer exists
```
(venv) (base) erfanarsala@MacBook-Pro-2 cti-sys % pip3 index versions psycopg-binary 
psycopg-binary (3.2.5)
Available versions: 3.2.5, 3.2.4, 3.1.19, 3.1.18, 3.1.17, 3.1.16, 3.1.15, 3.1.14, 3.1.13, 3.1.12
  INSTALLED: 3.2.4
  LATEST:    3.2.5
```
### Checklist
- [x] I've reviewed the contribution guide
- [x] I've confirmed the changes work on my machine
- [x] This pull request is ready for review

NickGuerrero commented:
Requests
* Ensure if they're primary email is changing, it matches the email they used to submit the form. This ensures we authentication based on their Google account.
* We can remove the fname, lname fields since they're unused. We'll have to work on logging later

Additional Notes
* I'll verify this with a new Google form before clearing the PR. I'll keep you updated on the status
* Contributing guide update: Library changes need to specified in the PR
* Models will be split into separate files, database models and application models will be distinguished

arsalaerfan commented:
> Requests
> 
> * Ensure if they're primary email is changing, it matches the email they used to submit the form. This ensures we authentication based on their Google account.
> * We can remove the fname, lname fields since they're unused. We'll have to work on logging later
> 
> Additional Notes
> 
> * I'll verify this with a new Google form before clearing the PR. I'll keep you updated on the status
> * Contributing guide update: Library changes need to specified in the PR
> * Models will be split into separate files, database models and application models will be distinguished

arsalaerfan commented:
> Requests
> 
> * Ensure if they're primary email is changing, it matches the email they used to submit the form. This ensures we authentication based on their Google account.
> * We can remove the fname, lname fields since they're unused. We'll have to work on logging later
> 
> Additional Notes
> 
> * I'll verify this with a new Google form before clearing the PR. I'll keep you updated on the status
> * Contributing guide update: Library changes need to specified in the PR
> * Models will be split into separate files, database models and application models will be distinguished

arsalaerfan commented:
> Requests
> 
> * Ensure if they're primary email is changing, it matches the email they used to submit the form. This ensures we authentication based on their Google account.
> * We can remove the fname, lname fields since they're unused. We'll have to work on logging later
> 
> Additional Notes
> 
> * I'll verify this with a new Google form before clearing the PR. I'll keep you updated on the status
> * Contributing guide update: Library changes need to specified in the PR
> * Models will be split into separate files, database models and application models will be distinguished

Updated PR to remove fname and lname and to ensure if they're primary email is changing, it matches the email they used to submit the form. Added note on Library change.

NickGuerrero commented on requirements.txt:
```diff
@@ -22,7 +22,7 @@ MarkupSafe==3.0.2
 packaging==24.2
 psycopg==3.2.3
-psycopg-binary==3.2.3
+psycopg-binary==3.2.4
```
As far as I'm aware, v3.2.3 still exists, per their [release notes](https://www.psycopg.org/psycopg3/docs/news.html#psycopg-3-2-4). It's a small iteration that doesn't affect the project or the PR, however it may be an issue on your end. Make sure both your Python and pip installations are up-to-date. This will not stop the PR, but I'm noting this here for future reference.

Additionally, it's safer to install the whole package (psycopg[binary,pool] is what I used), rather than incrementing. I can't tell if that's what you did, but python's psycopg staying at v3.2.3 is a bit concerning.

NickGuerrero replied:
As far as I'm aware, v3.2.3 still exists, per their [release notes](https://www.psycopg.org/psycopg3/docs/news.html#psycopg-3-2-4). It's a small iteration that doesn't affect the project or the PR, however it may be an issue on your end. Make sure both your Python and pip installations are up-to-date. This will not stop the PR, but I'm noting this here for future reference.

Additionally, it's safer to install the whole package (psycopg[binary,pool] is what I used), rather than incrementing. I can't tell if that's what you did, but python's psycopg staying at v3.2.3 is a bit concerning.

NickGuerrero commented on src/app/models/postgres/models.py:
```diff
@@ -148,4 +150,16 @@ class AccountabilityGroup(Base):
+    alt_emails: List[EmailStr] = []
+    primary_email: Optional[str] = None
+    remove_emails: List[EmailStr] = []
+    google_form_email: EmailStr
```
As mentioned in the review, this class should be moved to schema with the other pydantic schemas. Pydantic also doesn't need to be imported here. Note for future refactor.

NickGuerrero commented on src/main.py:
```diff
@@ -105,3 +107,106 @@ def create_application(
+def modify_alternate_emails(
+    request: AlternateEmailRequest,
+    db: Session = Depends(make_session),
+):
```
Functionality looks good. Although this is all under 1 endpoint, it may have been helpful to separate the functionalities (add new alternate email, remove alternate email, etc.) into different functions. It would make unit testing easier (single functions) and more clear from the endpoint testing (all the functions working together). Additionally, it would allow the HTTPExceptions to stand out more, since you currently have to follow the whole code logic to determine where an error would occur. I'm also not a fan of mid-loop continues, but I can allow it since they're very early in the loops here.

The exception handling at the bottom will be moved, but it's nice that you specified which errors you expect to run into. This section should also be reviewed once logging is more integrated with the project.

NickGuerrero commented on tests/test_main.py:
```diff
@@ -455,3 +461,222 @@ def test_courses_schema_accepts_valid_insert(self, real_mongo_db: pymongo.databa
+        })
+
+        assert response.status_code == 200
+        assert response.json() == {"status": 200}
```
The tests are fine, but how they're asserted is a problem. A flaw in this endpoint is that successful requests only return status 200, regardless of what actually changed. Since the success tests only check the status code, it's possible that a request with no errors could return 200, regardless of whether or not the operation was successfully completed. I can verify these tests by manually querying the database, but it's a potential blindspot in automated testing.

It's possible this issue is outside the scope of the current issue: We would need two separate modes of running the app (Development & Production), or we would need a different way of validating that the correct data modifications happened.

NickGuerrero reviewed (approved):
* Endpoint's functionality works like it should. Despite the somewhat changing requirements, everything works as intended in development. Some styling notes in the conversation.
* New schemas added look good, you also fixed one of the test endpoints. As noted, they belong in different files, but they should be moved soon.
* Fail tests look good. Success tests have a noticeable blind spot.
Overall, I'm satisfied with the work you put in here. I have some concerns noted in the comments - I recommend prioritizing the test code, since the blind spot is noticeable.

Note that I manually tested the endpoint. In making the connected Google Form, I realized that there could be many possible errors that have nothing to do with the code in this PR. The script code will need to be handled separately.
